File: Sources/PUPSISPortalApp/Core/GoogleCalendarClient.swift
import Foundation

/// A Google calendar the user can export into.
struct GoogleCalendar: Identifiable, Equatable {
    let id: String
    let summary: String
}

/// One event as Google's Calendar API wants it. Encodable so the export body is
/// built purely (and unit-tested) before it ever hits the network.
struct GoogleEvent: Encodable, Equatable {
    struct DateTime: Encodable, Equatable {
        let dateTime: String
        let timeZone: String
    }
    struct Extended: Encodable, Equatable {
        let `private`: [String: String]
    }
    let summary: String
    let location: String?
    let description: String?
    let start: DateTime
    let end: DateTime
    let recurrence: [String]
    let extendedProperties: Extended
}

/// Writes classes straight to Google Calendar over the REST API — the reliable
/// path when Google-via-Apple-Calendar (CalDAV) won't take recurring writes.
///
/// Mirrors `CalendarBridge.exportClasses`: it clears its own previous export
/// first (events tagged `pupsisportal=1` in `extendedProperties.private`) so a
/// re-export replaces rather than stacks, and leaves the user's other events
/// alone. Term status only, like the `.ics` export — a single repeating event
/// can't carry a one-week exception.
@MainActor
final class GoogleCalendarClient {
    /// The tag on every event this app writes, so a re-export removes only ours.
    static let tagKey = "pupsisportal"
    static let tagValue = "1"

    private let auth: GoogleAuth

    init(auth: GoogleAuth) {
        self.auth = auth
    }

    enum ClientError: LocalizedError {
        case http(String)
        var errorDescription: String? {
            switch self { case .http(let m): m }
        }
    }

    // MARK: Calendars

    func listCalendars() async throws -> [GoogleCalendar] {
        struct List: Decodable {
            let items: [Item]
            struct Item: Decodable { let id: String; let summary: String; let accessRole: String }
        }
        let url = URL(string: "https://www.googleapis.com/calendar/v3/users/me/calendarList")!
        let data = try await get(url)
        let list = try JSONDecoder().decode(List.self, from: data)
        // Only calendars the user can write to are useful as export targets.
        return list.items
            .filter { $0.accessRole == "owner" || $0.accessRole == "writer" }
            .map { GoogleCalendar(id: $0.id, summary: $0.summary) }
    }

    // MARK: Export

    /// Returns a short summary of what was written, so a silent no-op is
    /// impossible to mistake for success.
    @discardableResult
    func exportClasses(
        _ sessions: [ClassSession],
        weekStart: Date,
        until termEnd: Date,
        toCalendarID calendarID: String,
        status: (ClassSession) -> SessionStatus,
        calendar: Calendar = .current,
        timeZone: TimeZone = .current
    ) async throws -> String {
        let removed = try await clearExported(calendarID: calendarID)

        var written = 0
        for session in sessions {
            guard let event = Self.event(
                for: session, status: status(session),
                weekStart: weekStart, until: termEnd,
                calendar: calendar, timeZone: timeZone
            ) else { continue }
            try await insert(event, calendarID: calendarID)
            written += 1
        }

        let replaced = removed > 0 ? ", replacing \(removed)" : ""
        return "Exported \(written) class\(written == 1 ? "" : "es") to Google\(replaced)."
    }

    /// The event body for a class in its term status, or nil when it shouldn't be
    /// written (vacant). Pure, so the summary/RRULE/tag are unit-tested without
    /// the network. Reuses `CalendarBridge.ClassExport.plan` so Google and the
    /// EventKit/.ics exports agree on skip/online rules.
    static func event(
        for session: ClassSession,
        status: SessionStatus,
        weekStart: Date,
        until termEnd: Date,
        calendar: Calendar = .current,
        timeZone: TimeZone = .current
    ) -> GoogleEvent? {
        guard case let .event(titleSuffix, location) = CalendarBridge.ClassExport.plan(for: status) else {
            return nil
        }

        let midnight = session.day.date(inWeekStarting: weekStart, calendar: calendar)
        guard let start = calendar.date(byAdding: .minute, value: session.start, to: midnight),
              let end = calendar.date(byAdding: .minute, value: session.end, to: midnight)
        else { return nil }

        let last = ClassRecurrence.lastMoment(of: termEnd, calendar: calendar)
        // Start carries a timezone, so RFC-5545 wants the RRULE UNTIL in UTC.
        let rule = "RRULE:\(ClassRecurrence.rrule(day: session.day, until: ClassRecurrence.utc(last)))"

        let iso = ISO8601DateFormatter()
        iso.timeZone = timeZone
        iso.formatOptions = [.withInternetDateTime]
        let tz = timeZone.identifier

        return GoogleEvent(
            summary: session.subjectCode + (titleSuffix ?? ""),
            location: location,
            description: session.faculty.isEmpty ? session.description : "\(session.description)\n\(session.faculty)",
            start: .init(dateTime: iso.string(from: start), timeZone: tz),
            end: .init(dateTime: iso.string(from: end), timeZone: tz),
            recurrence: [rule],
            extendedProperties: .init(private: [tagKey: tagValue])
        )
    }

    // MARK: Detail

    private func clearExported(calendarID: String) async throws -> Int {
        struct List: Decodable {
            let items: [Item]
            struct Item: Decodable { let id: String }
        }
        var comps = URLComponents(string: "https://www.googleapis.com/calendar/v3/calendars/\(encode(calendarID))/events")!
        comps.queryItems = [
            .init(name: "privateExtendedProperty", value: "\(Self.tagKey)=\(Self.tagValue)"),
            .init(name: "maxResults", value: "2500"),
            // Recurring events return as the master here, so deleting it drops the
            // whole series in one call.
            .init(name: "singleEvents", value: "false"),
        ]
        let data = try await get(comps.url!)
        let list = try JSONDecoder().decode(List.self, from: data)

        for item in list.items {
            let url = URL(string: "https://www.googleapis.com/calendar/v3/calendars/\(encode(calendarID))/events/\(encode(item.id))")!
            try await delete(url)
        }
        return list.items.count
    }

    private func insert(_ event: GoogleEvent, calendarID: String) async throws {
        let url = URL(string: "https://www.googleapis.com/calendar/v3/calendars/\(encode(calendarID))/events")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(event)
        _ = try await send(request)
    }

    // MARK: HTTP

    private func get(_ url: URL) async throws -> Data {
        try await send(URLRequest(url: url))
    }

    private func delete(_ url: URL) async throws {
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        _ = try await send(request)
    }

    private func send(_ request: URLRequest) async throws -> Data {
        var request = request
        let token = try await auth.validAccessToken()
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw ClientError.http("No response from Google.") }
        // A delete of an already-gone event is fine.
        guard (200..<300).contains(http.statusCode) || http.statusCode == 410 else {
            throw ClientError.http(Self.googleError(data, status: http.statusCode))
        }
        return data
    }

    private func encode(_ component: String) -> String {
        component.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? component
    }

    private static func googleError(_ data: Data, status: Int) -> String {
        struct E: Decodable { struct Err: Decodable { let message: String }; let error: Err? }
        if let e = try? JSONDecoder().decode(E.self, from: data), let m = e.error?.message {
            return "Google: \(m)"
        }
        return "Google request failed (\(status))."
    }
}

File: windows/PUPSISPortal.Core/GoogleCalendarClient.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Web;

namespace PUPSISPortal.Core;

/// <summary>
/// A Google calendar the user can export into.
/// </summary>
public class GoogleCalendar
{
    [JsonPropertyName("id")]
    public required string Id { get; set; }

    [JsonPropertyName("summary")]
    public required string Summary { get; set; }
}

/// <summary>
/// One event as Google's Calendar API wants it. Serializable so the export body is
/// built purely (and unit-tested) before it ever hits the network.
/// </summary>
public class GoogleEvent
{
    [JsonPropertyName("summary")]
    public required string Summary { get; set; }

    [JsonPropertyName("location")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Location { get; set; }

    [JsonPropertyName("description")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Description { get; set; }

    [JsonPropertyName("start")]
    public required GoogleEventDateTime Start { get; set; }

    [JsonPropertyName("end")]
    public required GoogleEventDateTime End { get; set; }

    [JsonPropertyName("recurrence")]
    public required List<string> Recurrence { get; set; }

    [JsonPropertyName("extendedProperties")]
    public required GoogleEventExtendedProperties ExtendedProperties { get; set; }
}

public class GoogleEventDateTime
{
    [JsonPropertyName("dateTime")]
    public required string DateTime { get; set; }

    [JsonPropertyName("timeZone")]
    public required string TimeZone { get; set; }
}

public class GoogleEventExtendedProperties
{
    [JsonPropertyName("private")]
    public required Dictionary<string, string> Private { get; set; }
}

/// <summary>
/// Writes classes straight to Google Calendar over the REST API — the reliable
/// path when Google-via-system-Calendar (CalDAV) won't take recurring writes.
///
/// Mirrors the calendar export: it clears its own previous export first (events
/// tagged `pupsisportal=1` in `extendedProperties.private`) so a re-export replaces
/// rather than stacks, and leaves the user's other events alone. Term status only,
/// like the `.ics` export — a single repeating event can't carry a one-week exception.
/// </summary>
public class GoogleCalendarClient
{
    /// <summary>
    /// The tag on every event this app writes, so a re-export removes only ours.
    /// </summary>
    public const string TagKey = "pupsisportal";
    public const string TagValue = "1";

    private readonly Func<Task<string>> _getAccessToken;
    private static readonly HttpClient HttpClient = new();

    public GoogleCalendarClient(Func<Task<string>> getAccessToken)
    {
        _getAccessToken = getAccessToken;
    }

    public class ClientException : Exception
    {
        public ClientException(string message) : base(message) { }
    }

    // MARK: Calendars

    public async Task<List<GoogleCalendar>> ListCalendarsAsync()
    {
        const string url = "https://www.googleapis.com/calendar/v3/users/me/calendarList";
        var data = await GetAsync(url);

        var list = JsonSerializer.Deserialize<CalendarList>(data);
        if (list?.Items == null)
            return new List<GoogleCalendar>();

        // Only calendars the user can write to are useful as export targets.
        return list.Items
            .Where(item => item.AccessRole == "owner" || item.AccessRole == "writer")
            .Select(item => new GoogleCalendar { Id = item.Id, Summary = item.Summary })
            .ToList();
    }

    // MARK: Export

    /// <summary>
    /// Returns a short summary of what was written, so a silent no-op is
    /// impossible to mistake for success.
    /// </summary>
    public async Task<string> ExportClassesAsync(
        List<ClassSession> sessions,
        DateTime weekStart,
        DateTime termEnd,
        string calendarId,
        Func<ClassSession, SessionStatus> statusResolver)
    {
        var removed = await ClearExportedAsync(calendarId);

        var written = 0;
        foreach (var session in sessions)
        {
            var googleEvent = Event(session, statusResolver(session), weekStart, termEnd);
            if (googleEvent == null)
                continue;

            await InsertAsync(googleEvent, calendarId);
            written++;
        }

        var replaced = removed > 0 ? $", replacing {removed}" : "";
        return $"Exported {written} class{(written == 1 ? "" : "es")} to Google{replaced}.";
    }

    /// <summary>
    /// The event body for a class in its term status, or null when it shouldn't be
    /// written (vacant). Pure, so the summary/RRULE/tag are unit-tested without
    /// the network. Accepts timezone for testing (defaults to system local).
    /// </summary>
    public static GoogleEvent? Event(
        ClassSession session,
        SessionStatus status,
        DateTime weekStart,
        DateTime termEnd,
        TimeZoneInfo? timeZone = null)
    {
        var plan = ClassExportPlan(status);
        if (plan is null)
            return null;

        var (titleSuffix, location) = plan.Value;

        timeZone ??= TimeZoneInfo.Local;
        var tzId = timeZone.Id;

        var midnight = session.Day.DateInWeekStarting(weekStart);
        var start = midnight.AddMinutes(session.Start);
        var end = midnight.AddMinutes(session.End);

        var last = ClassRecurrence.LastMoment(termEnd);

        // Start carries a timezone, so RFC-5545 wants the RRULE UNTIL in UTC.
        var rule = $"RRULE:{ClassRecurrence.RRule(session.Day, ClassRecurrence.Utc(last))}";

        // Format as ISO8601 without timezone indicator; the timezone is in the separate field
        var iso = start.ToString("yyyy-MM-ddTHH:mm:ss");
        var endIso = end.ToString("yyyy-MM-ddTHH:mm:ss");

        return new GoogleEvent
        {
            Summary = session.SubjectCode + (titleSuffix ?? ""),
            Location = location,
            Description = string.IsNullOrEmpty(session.Faculty)
                ? session.Description
                : $"{session.Description}\n{session.Faculty}",
            Start = new GoogleEventDateTime { DateTime = iso, TimeZone = tzId },
            End = new GoogleEventDateTime { DateTime = endIso, TimeZone = tzId },
            Recurrence = new List<string> { rule },
            ExtendedProperties = new GoogleEventExtendedProperties
            {
                Private = new Dictionary<string, string> { { TagKey, TagValue } }
            }
        };
    }

    // MARK: Detail

    private async Task<int> ClearExportedAsync(string calendarId)
    {
        var comps = $"https://www.googleapis.com/calendar/v3/calendars/{UrlEncode(calendarId)}/events?" +
                    $"privateExtendedProperty={UrlEncode($"{TagKey}={TagValue}")}&" +
                    "maxResults=2500&" +
                    "singleEvents=false";

        var data = await GetAsync(comps);
        var list = JsonSerializer.Deserialize<EventList>(data);

        var count = 0;
        if (list?.Items != null)
        {
            foreach (var item in list.Items)
            {
                var url = $"https://www.googleapis.com/calendar/v3/calendars/{UrlEncode(calendarId)}/events/{UrlEncode(item.Id)}";
                try
                {
                    await DeleteAsync(url);
                    count++;
                }
                catch
                {
                    // Ignore errors on individual deletes
                }
            }
        }

        return count;
    }

    private async Task InsertAsync(GoogleEvent googleEvent, string calendarId)
    {
        var url = $"https://www.googleapis.com/calendar/v3/calendars/{UrlEncode(calendarId)}/events";
        var json = JsonSerializer.Serialize(googleEvent);
        var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");

        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = content
        };

        await SendAsync(request);
    }

    // MARK: HTTP

    private async Task<string> GetAsync(string url)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        return await SendAsync(request);
    }

    private async Task DeleteAsync(string url)
    {
        using var request = new HttpRequestMessage(HttpMethod.Delete, url);
        await SendAsync(request);
    }

    private async Task<string> SendAsync(HttpRequestMessage request)
    {
        var token = await _getAccessToken();
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);

        var response = await HttpClient.SendAsync(request);

        // A delete of an already-gone event is fine.
        if (!response.IsSuccessStatusCode && response.StatusCode != System.Net.HttpStatusCode.Gone)
        {
            var body = await response.Content.ReadAsStringAsync();
            throw new ClientException(GoogleError(body, (int)response.StatusCode));
        }

        return await response.Content.ReadAsStringAsync();
    }

    private static string UrlEncode(string component)
    {
        return HttpUtility.UrlEncode(component);
    }

    private static string GoogleError(string data, int status)
    {
        try
        {
            var errorResponse = JsonSerializer.Deserialize<GoogleErrorResponse>(data);
            if (errorResponse?.Error?.Message != null)
            {
                return $"Google: {errorResponse.Error.Message}";
            }
        }
        catch
        {
            // Ignore parsing errors
        }

        return $"Google request failed ({status}).";
    }

    // MARK: JSON Models

    private class CalendarList
    {
        [JsonPropertyName("items")]
        public List<CalendarItem>? Items { get; set; }
    }

    private class CalendarItem
    {
        [JsonPropertyName("id")]
        public required string Id { get; set; }

        [JsonPropertyName("summary")]
        public required string Summary { get; set; }

        [JsonPropertyName("accessRole")]
        public required string AccessRole { get; set; }
    }

    private class EventList
    {
        [JsonPropertyName("items")]
        public List<EventItem>? Items { get; set; }
    }

    private class EventItem
    {
        [JsonPropertyName("id")]
        public required string Id { get; set; }
    }

    private class GoogleErrorResponse
    {
        [JsonPropertyName("error")]
        public ErrorDetail? Error { get; set; }
    }

    private class ErrorDetail
    {
        [JsonPropertyName("message")]
        public string? Message { get; set; }
    }

    // MARK: Export plan helper

    private static (string? titleSuffix, string? location)? ClassExportPlan(SessionStatus sessionStatus)
    {
        return sessionStatus switch
        {
            SessionStatus.Regular => ("", null),
            SessionStatus.Online => (" (Online)", "Online"),
            SessionStatus.Vacant => null,
            _ => null,
        };
    }
}
